frontend/dockerfile/dockerfile_test.go

```diff
@@ -36,48 +36,125 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestIntegration(t *testing.T) {
-	integration.Run(t, []integration.Test{
-		testNoSnapshotLeak,
-		testCmdShell,
-		testGlobalArg,
-		testDockerfileDirs,
-		testDockerfileInvalidCommand,
-		testDockerfileADDFromURL,
-		testDockerfileAddArchive,
-		testDockerfileScratchConfig,
-		testExportedHistory,
-		testExposeExpansion,
-		testUser,
-		testDockerignore,
-		testDockerignoreInvalid,
-		testDockerfileFromGit,
-		testCopyChown,
-		testCopyWildcards,
-		testCopyOverrideFiles,
-		testMultiStageImplicitFrom,
-		testCopyVarSubstitution,
-		testMultiStageCaseInsensitive,
-		testLabels,
-		testCacheImportExport,
-		testReproducibleIDs,
-		testImportExportReproducibleIDs,
-		testNoCache,
-		testDockerfileFromHTTP,
-		testBuiltinArgs,
-		testPullScratch,
-		testSymlinkDestination,
-		testHTTPDockerfile,
-		testNoSnapshotLeak,
-		testCopySymlinks,
-		testContextChangeDirToFile,
-		testPlatformArgsImplicit,
-		testPlatformArgsExplicit,
-		testExportMultiPlatform,
+type Mode string
+
+var (
+	ModeBuiltin       = Mode("builtin")
+	ModeMasterGateway = Mode("master-gateway")
+)
+
+type SolverFn func(context.Context, client.SolveOpt, chan *client.SolveStatus) (*client.SolveResponse, error)
+type frontend struct {
+	t    *testing.T
+	mode Mode
+}
+
+func (f *frontend) modeIs(m Mode) bool {
+	return f.mode == m
+}
+
+func (f *frontend) solver(c *client.Client) SolverFn {
+	switch f.mode {
+	case ModeBuiltin:
+		return func(ctx context.Context, opt client.SolveOpt, statusChan chan *client.SolveStatus) (*client.SolveResponse, error) {
+			opt.Frontend = "dockerfile.v0"
+			return c.Solve(ctx, nil, opt, statusChan)
+		}
+	case ModeMasterGateway:
+		return func(ctx context.Context, opt client.SolveOpt, statusChan chan *client.SolveStatus) (*client.SolveResponse, error) {
+			opt.Frontend = "gateway.v0"
+			if opt.FrontendAttrs == nil {
+				opt.FrontendAttrs = make(map[string]string)
+			}
+			opt.FrontendAttrs["source"] = "tonistiigi/dockerfile"
```

Comment:
`tonistiigi/dockerfile:master`

Reply:
Done, I had misread https://hub.docker.com/r/tonistiigi/dockerfile/tags/. I will also revisit "dockerfile: skip some tests which appear to fail in gateway mode" since this may have changed the results.

Reply:
> will also revisit "dockerfile: skip some tests which appear to fail in gateway mode"

All apart from `TestDockerfileInvalidCommand` now pass and I can see the issue with that one so I updated the skip reason and commit message to reflect that.

It does imply that image skew is a real problem though.